### PR TITLE
Frames decoder gpu without index

### PR DESCRIPTION
**Pull request by awolant**
- adds an ability to decode the video on the GPU without
  building the internal index

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
- adds an ability to decode the video on the GPU without
  building the internal index

### Affected modules and functionalities:
- frames_decoder_gpu

### Key points relevant for the review:
- NA

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
    - FramesDecoderGpuTest.VariableFrameRateNoIndex
    - FramesDecoderGpuTest.VariableFrameRateHevcNoIndex 
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3038

### CODE REVIEW

**Comment by awolant** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 1]
```diff
@@ -36,7 +36,11 @@ int process_video_sequence(void *user_data, CUVIDEOFORMAT *video_format) {
 int process_picture_decode(void *user_data, CUVIDPICPARAMS *picture_params) {
   FramesDecoderGpu *frames_decoder = static_cast<FramesDecoderGpu*>(user_data);
 
-  return frames_decoder->ProcessPictureDecode(user_data, picture_params);
+  if (frames_decoder->HasIndex()) {
+    return frames_decoder->ProcessPictureDecodeWithIndex(user_data, picture_params);
+  }
+
+  return frames_decoder->ProcessPictureDecodeWithoutIndex(user_data, picture_params);
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:43]: In this case I think that if/else is more verbose about one of the options, as it is not early exit after all.
```suggestion
  if (frames_decoder->HasIndex()) {
    return frames_decoder->ProcessPictureDecodeWithIndex(user_data, picture_params);
  } else {
    return frames_decoder->ProcessPictureDecodeWithoutIndex(user_data, picture_params);
  }
```

Comment (awolant) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:43]: Done

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 2]
```diff
@@ -288,6 +353,126 @@ bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
...
+      frame_returned_ = false;
+      CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser, packet));
+    } else {
+      SendLastPacket();
+      more_frames_to_decode_ = false;
+    }
+  }
+
+  int frame_to_return_index = -1;
+  for (int i = 0; i < num_decode_surfaces_; ++i) {
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:396]:
```suggestion
  for (int i = 0; i < frame_buffer_.size(); ++i) {
```

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 3]
```diff
@@ -216,31 +221,91 @@ int FramesDecoderGpu::ProcessPictureDecode(void *user_data, CUVIDPICPARAMS *pict
   return 1;
 }
 
+int FramesDecoderGpu::ProcessPictureDecodeWithoutIndex(
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:224]: I think you can extract common part from here and `ProcessPictureDecodeWithIndex`.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 4]
```diff
@@ -216,31 +221,91 @@ int FramesDecoderGpu::ProcessPictureDecode(void *user_data, CUVIDPICPARAMS *pict
...
+  videoProcessingParameters.second_field = 1;
+  videoProcessingParameters.top_field_first = picture_params->bottom_field_flag ^ 1;
+  videoProcessingParameters.unpaired_field = 0;
+  videoProcessingParameters.output_stream = stream_;
+
+  uint8_t *frame_output = nullptr;
+
+  // Take pts of the currently decoded frame
+  int current_pts = piped_pts_.front();
+  piped_pts_.pop();
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:246]: I would add `LOG_LINE` with the decoded frame pts for the further debugging.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 5]
```diff
@@ -288,6 +353,126 @@ bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
...
+      packet->timestamp = filtered_packet_->pts;
+
+      // Send packet to the nv decoder
+      frame_returned_ = false;
+      CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser, packet));
+    } else {
+      SendLastPacket();
+      more_frames_to_decode_ = false;
+    }
+  }
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:458]: Seems similar to L361.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 6]
```diff
@@ -320,8 +505,28 @@ BufferedFrame& FramesDecoderGpu::FindEmptySlot() {
   DALI_FAIL("Could not find empty slot in the frame buffer");
 }
 
+bool FramesDecoderGpu::HasEmptySlot() const {
+  for (auto &frame : frame_buffer_) {
```
Comment (JanuszL) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:509]: Maybe there should be a counter for the empty spots? Just a thought.

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:406]: On the other hand it may be not worth it.

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder.h [hunk 7]
```diff
@@ -191,6 +191,13 @@ class DLL_PUBLIC FramesDecoder {
    */
   int NextFrameIdx() { return next_frame_idx_; }
 
+  /**
+ * @brief Returns true if the index was build.
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder.h:195]:
```suggestion
 * @brief Returns true if the index was built.
```

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 8]
```diff
@@ -134,14 +134,15 @@ FramesDecoderGpu::FramesDecoderGpu(const std::string &filename, cudaStream_t str
...
+  const char *memory_file, int memory_file_size, cudaStream_t stream, bool build_index) :
+  FramesDecoder(memory_file, memory_file_size, build_index),
   frame_buffer_(num_decode_surfaces_),
   stream_(stream) {
   InitGpuDecoder();
 }
 
-int FramesDecoderGpu::ProcessPictureDecode(void *user_data, CUVIDPICPARAMS *picture_params) {
+int FramesDecoderGpu::ProcessPictureDecode(
+  void *user_data, CUVIDPICPARAMS *picture_params) {
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:145]: Why? It doesn't need line breaks to fit in our line length limit.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 9]
```diff
@@ -288,6 +266,110 @@ bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
...
+  // No more frames in the file
+  if (next_frame_idx_ == -1) {
+    return false;
+  }
+
+  if (HasIndex()) {
+    return ReadNextFrameWithIndex(data, copy_to_output);
+  }
+
+  return ReadNextFrameWithoutIndex(data, copy_to_output);
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:370]: Nitpick - for symmetry
```suggestion
  if (HasIndex()) {
    return ReadNextFrameWithIndex(data, copy_to_output);
  } else {
    return ReadNextFrameWithoutIndex(data, copy_to_output);
  }
```

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 10]
```diff
@@ -320,8 +402,28 @@ BufferedFrame& FramesDecoderGpu::FindEmptySlot() {
...
+  for (auto &frame : frame_buffer_) {
+    if (frame.pts_ == -1) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool FramesDecoderGpu::EmptyBuffer() const {
+  for (auto &frame_ : frame_buffer_) {
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:415]:
```suggestion
  for (auto &frame : frame_buffer_) {
```

File: dali/operators/reader/loader/video/frames_decoder.h (hunk 7 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder.h:195]: Done

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 8 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:145]: Done

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 9 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:370]: Done

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 10 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:415]: Done

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video/frames_decoder.h [hunk 11]
```diff
@@ -90,107 +90,114 @@ struct MemoryVideoFile {
...
+   *
    * @note This constructor assumes that the `memory_file` and
    * `memory_file_size` arguments cover the entire video file, including the header.
    */
   FramesDecoder(const char *memory_file, int memory_file_size, bool build_index = true);
 
   /**
    * @brief Number of frames in the video. It returns 0, if this information is unavailable.
-   * 
+   *
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder.h:121]: When you touch these lines - they should just go away if we don't have anything after "brief".

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder.h]: The IDE trimmed the trailing whitespaces. Nevertheless, removed.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 12]
```diff
@@ -222,59 +225,33 @@ void FramesDecoderGpu::SeekFrame(int frame_id) {
...
-  }
-
+bool FramesDecoderGpu::ReadNextFrameWithIndex(uint8_t *data, bool copy_to_output) {
   // Check if requested frame was buffered earlier
   for (auto &frame : frame_buffer_) {
-    if (frame.pts_ == Index(next_frame_idx_).pts) {
-      if (copy_to_output) {
-        copyD2D(data, frame.frame_.data(), FrameSize());
+    if (frame.pts_ != -1) {
+      if (frame.pts_ == Index(next_frame_idx_).pts) {
```
Comment (stiepan) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:232]: nitpick: Wouldn't it be slightly more readable with just `&&`?

Btw. Is this enforced by the changes in this PR or more of a "hardnening" the check unrealted to the changes?

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 13]
```diff
@@ -288,6 +265,110 @@ bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
...
+  while (HasEmptySlot() && more_frames_to_decode_) {
+    if (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+      if (!SendFrameToParser()) {
+        continue;
+      }
+    } else {
+      SendLastPacket();
+      more_frames_to_decode_ = false;
+    }
+  }
```
Comment (stiepan) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:354]: 1. Could it be a separate private method if that is called twice: as "a prolog" and "epilogue" of this method? 
2. Why do we need it twice? Is this just to fill the one gap created by `frame_buffer_[frame_to_return_index].pts_ = -1;`?

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:354]: I guess we can remove the epilog.

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc (hunk 12 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc:232]: Fixed

**Review by stiepan — APPROVED** [review]

File: dali/operators/reader/loader/video/frames_decoder_gpu.h [hunk 14]
```diff
@@ -108,11 +114,21 @@ class DLL_PUBLIC FramesDecoderGpu : public FramesDecoder {
 
   BufferedFrame& FindEmptySlot();
 
+  bool HasEmptySlot() const;
+
+  bool EmptyBuffer() const;
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.h:119]: Reads like "empty the buffer" == "make it empty".
```suggestion
  bool BufferEmpty() const;
```
or
```suggestion
  bool IsBufferEmpty() const;
```

File: dali/operators/reader/loader/video/frames_decoder_gpu.cc [hunk 15]
```diff
@@ -320,8 +388,28 @@ BufferedFrame& FramesDecoderGpu::FindEmptySlot() {
...
+bool FramesDecoderGpu::HasEmptySlot() const {
+  for (auto &frame : frame_buffer_) {
+    if (frame.pts_ == -1) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool FramesDecoderGpu::EmptyBuffer() const {
```
Comment (mzient) [inline dali/operators/reader/loader/video/frames_decoder_gpu.cc:400]:
```suggestion
bool FramesDecoderGpu::BufferEmpty() const {
```
or
```suggestion
bool FramesDecoderGpu::IsBufferEmpty() const {
```

Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.cc]: Fixed

File: dali/operators/reader/loader/video/frames_decoder_gpu.h (hunk 14 above)
Comment (JanuszL) [reply dali/operators/reader/loader/video/frames_decoder_gpu.h]: Fixed

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Pull request merged**